From: Jens Axboe <axboe@fb.com>
Date: Fri, 9 Dec 2016 13:08:35 -0700
Subject: blk-stat: fix a few cases of missing batch flushing
References: bsc#1020989,FATE#322379
Git-commit: 7cd54aa8438947602cf68eda1db327822b9b8e6b
Patch-Mainline: v4.10-rc1

Everytime we need to read ->nr_samples, we should have flushed
the batch first. The non-mq read path also needs to flush the
batch.

Signed-off-by: Jens Axboe <axboe@fb.com>
Signed-off-by: Hannes Reinecke <hare@suse.de>
---
 block/blk-stat.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/block/blk-stat.c b/block/blk-stat.c
index 4d01185..9b43efb 100644
--- a/block/blk-stat.c
+++ b/block/blk-stat.c
@@ -64,6 +64,9 @@ static void blk_mq_stat_get(struct request_queue *q, struct blk_rq_stat *dst)
 
 		queue_for_each_hw_ctx(q, hctx, i) {
 			hctx_for_each_ctx(hctx, ctx, j) {
+				blk_stat_flush_batch(&ctx->stat[BLK_STAT_READ]);
+				blk_stat_flush_batch(&ctx->stat[BLK_STAT_WRITE]);
+
 				if (!ctx->stat[BLK_STAT_READ].nr_samples &&
 				    !ctx->stat[BLK_STAT_WRITE].nr_samples)
 					continue;
@@ -111,6 +114,8 @@ void blk_queue_stat_get(struct request_queue *q, struct blk_rq_stat *dst)
 	if (q->mq_ops)
 		blk_mq_stat_get(q, dst);
 	else {
+		blk_stat_flush_batch(&q->rq_stats[BLK_STAT_READ]);
+		blk_stat_flush_batch(&q->rq_stats[BLK_STAT_WRITE]);
 		memcpy(&dst[BLK_STAT_READ], &q->rq_stats[BLK_STAT_READ],
 				sizeof(struct blk_rq_stat));
 		memcpy(&dst[BLK_STAT_WRITE], &q->rq_stats[BLK_STAT_WRITE],
@@ -128,6 +133,9 @@ void blk_hctx_stat_get(struct blk_mq_hw_ctx *hctx, struct blk_rq_stat *dst)
 		uint64_t newest = 0;
 
 		hctx_for_each_ctx(hctx, ctx, i) {
+			blk_stat_flush_batch(&ctx->stat[BLK_STAT_READ]);
+			blk_stat_flush_batch(&ctx->stat[BLK_STAT_WRITE]);
+
 			if (!ctx->stat[BLK_STAT_READ].nr_samples &&
 			    !ctx->stat[BLK_STAT_WRITE].nr_samples)
 				continue;
-- 
1.8.5.6

